app/models/metadata.rb

```diff
@@ -0,0 +1,130 @@
+class Metadata
+  def initialize(metadata_file:, collection:)
+    @rowset_errors = []
+    @new_metadata = []
+    @canonical_metadata = []
+    @metadata_file = metadata_file
+    @collection = collection
+  end
+
+  def process_csv
+    begin
+      csv = CSV.open(@metadata_file)
+      headers = csv.shift
+    rescue CSV::MalformedCSVError
+      csv = CSV.open(@metadata_file, :encoding => "ISO8859-1")
+      headers = csv.shift
+    end
+
+    rows = CSV.parse(@metadata_file).map { |a| Hash[ headers.zip(a) ] }
+    rows.shift
+
+    # process rows.
+    rows.each do |row|
+      if row.include?('work_id')
+        row.each do |r|
+          @new_metadata << { label: r[0],  value: r[1] }
+        end
+
+        begin
+          work = Work.find(row['work_id'].to_i)
+          work.update(original_metadata: @new_metadata.to_json)
+          save_canonical_metadata(work)
+
+          unless @collection.works.include?(work)
+            @rowset_errors << { error: "No work with ID #{row['work_id']} is in collection #{@collection.title}",
+                                work_id: row['work_id'],
+                                title: row['title'] }
+          end
+        rescue ActiveRecord::RecordNotFound
+          @rowset_errors << { error: "No work exists with ID #{row['work_id']}",
+                              work_id: row['work_id'],
+                              title: row['title'] }
+
+          # write the error.csv to the filesystem.
+          output_file(@rowset_errors)
+        end
+      elsif row.include?('filename')
+        work = Work.where(uploaded_filename: row['filename']).first
+
+        if work.nil?
+          @rowset_errors << { error: "No work exists with filename #{row['filename']}" }
+          output_file(@rowset_errors)
+        else
+          row.each do |k, v|
+            encval = v.force_encoding('ISO-8859-1') unless v.nil?
+            @new_metadata << { label: k, value: encval }
+          end
+
+          work.update(original_metadata: @new_metadata.to_json)
+          save_canonical_metadata(work)
+        end
+      end
+    end
+
+    result = { content: rows, errors: @rowset_errors }
+    result
+  end
+
+  def save_canonical_metadata(work)
+    canonical_metadata = work.collection.canonical_metadata
+
+    if canonical_metadata.empty?
+      @new_metadata.each do |m|
+        @canonical_metadata << { m[:label] => 1 }
+      end
+    else
+      json_metadata = JSON.parse(work.collection.canonical_metadata)
+
+      unless json_metadata.empty?
+        json_metadata.each do |c|
+          c.each do |k, v|
+            c[k] = v + 1
```

Comment:
It occurs to me that this will increment the metadata coverage count every time a work is updated if we do this in an `after_save` callback.  Other options include doing this in an `after_create` call-back or only when the metadata field has been updated.